### DISCUSSION TITLE
Page Transitions/Animations Using Framer Motion

### DISCUSSION THREAD

**Discussion by salemlf** [TanStack/router#823] [Q&A] [answered]
I plan on migrating my app to Tanstack Router, but I wanted to know what implementing page transitions with Framer Motion would look like? React Router v6 makes animating page transitions kind of a pain compared to v5, and seemingly impossible in the case of nested routes, which is one of many reasons I want to migrate.

I'm currently using React Router v6 with `AnimatePresence` and an outlet. Wondering if I'd use a similar approach in Tanstack Router, and I'd appreciate examples as I've been unable to find any in the docs so far. 

Thanks!

**Comment by AltwargEvan** [comment]
This article talks about how to do it with NextJs.  https://blog.stackademic.com/how-to-perfect-slide-in-and-slide-out-page-transitions-in-next-js-with-framer-motion-67a2f320762

This should work the same way with Tanstack. In the PageTransitionLayout component, you will need to use useRouter from tanstack. The key prop for motion.div should be router.state.location.pathname I think

**Reply by egor-xyz** [reply]
How to do not in nextJS? Just in react

**Reply by Ashilex** [reply]
I can enter animate but not exit...
I set a key on motion component but nothing.

I think `AnimatePresence` component should wrap around the route component we want to exit. If we follow the linked guide, Router seems to remove from DOM tree the `AnimatePresence` which in turn is not able to detect route unmounting.
If you wrap `Outlet` component you can verify this and see how the exit animation is applied, to the new route though.

**Reply by egor-xyz** [reply]
It worked with react-router-dom when I added location and key. Is it possible to do the same with the tanstack router?
```jsx
 <Routes
        location={location}
        key={location.pathname}
      >
```

**Comment by fauxparse** [comment, marked as answer]
Okay, I think I've cracked this! The trick seems to be keeping hold of the old router context and using that when the component is on the way out. Thankfully, TSR exposes all the context stuff you need to make this work:

```typescript
const AnimatedOutlet = forwardRef<HTMLDivElement>((_, ref) => {
  const RouterContext = getRouterContext();

  const routerContext = useContext(RouterContext);

  const renderedContext = useRef(routerContext);

  const isPresent = useIsPresent();

  if (isPresent) {
    renderedContext.current = cloneDeep(routerContext);
  }

  return (
    <motion.div ref={ref} {...transitionProps}>
      <RouterContext.Provider value={renderedContext.current}>
        <Outlet />
      </RouterContext.Provider>
    </motion.div>
  );
});
```

Then in your parent component:

```typescript
const Root = () => {
  const matches = useMatches();
  const match = useMatch({ strict: false });
  const nextMatchIndex = matches.findIndex((d) => d.id === match.id) + 1;
  const nextMatch = matches[nextMatchIndex];

  return (
    <main>
      <AnimatePresence mode="popLayout">
        <AnimatedOutlet key={nextMatch.id} />
      </AnimatePresence>
    </main>
  );
};
```

I haven't tested this super extensively, but initial results seem promising.

Notes:
* I haven't included the `transitionProps` here, but it's your standard Framer Motion stuff like `initial`, `animate`, and `exit`. Do whatever.
* This uses `deepClone` from `lodash` because I was in a hurry and had `lodash` in the project anyway, but you can substitute any deep clone function. The important part is that it MUST clone `context.store.state`: just saving a reference to the old context won't cut it.
* This probably won't work with nested (animated) routes without some extra finesse (but I might be wrong!)

Would love to know if anyone uses this, and how you get on!

**Reply by fauxparse** [reply]
Update: I've been playing with this more and I don't think the above version plays nicely with params and nested routes. Here's my current `AnimatedOutlet` component:

```typescript
import { getRouterContext, Outlet, useMatches } from '@tanstack/react-router';
import { useIsPresent, motion, MotionProps, Variants } from 'framer-motion';
import { cloneDeep } from 'lodash-es';
import { forwardRef, useContext, useRef } from 'react';
import { Direction } from './types';

type AnimatedOutletProps = MotionProps & {
  direction?: Direction;
};

const OFFSET: Record<Direction, 1 | -1> = {
  left: 1,
  right: -1,
  up: 1,
  down: -1,
};

const AXIS: Record<Direction, 'x' | 'y'> = {
  left: 'x',
  right: 'x',
  up: 'y',
  down: 'y',
};

export const RouteTransitionVariants: Variants = {
  initial: (direction: Direction = 'left') => ({
    [AXIS[direction]]: `${OFFSET[direction] * 100}vw`,
    opacity: 0,
  }),
  animate: (direction: Direction = 'left') => ({
    [AXIS[direction]]: 0,
    opacity: 1,
  }),
  exit: (direction: Direction = 'left') => ({
    [AXIS[direction]]: `${OFFSET[direction] * -100}vw`,
    opacity: 0,
  }),
};

export const TransitionProps = {
  variants: RouteTransitionVariants,
  initial: 'initial',
  animate: 'animate',
  exit: 'exit',
  transition: {
    type: 'spring',
    bounce: 0.2,
  },
  style: {
    display: 'grid',
    alignSelf: 'stretch',
    justifySelf: 'stretch',
  },
} as const;

const AnimatedOutlet = forwardRef<HTMLDivElement, AnimatedOutletProps>(
  ({ direction, ...props }, ref) => {
    const isPresent = useIsPresent();

    const matches = useMatches();
    const prevMatches = useRef(matches);

    const RouterContext = getRouterContext();
    const routerContext = useContext(RouterContext);

    let renderedContext = routerContext;

    if (isPresent) {
      prevMatches.current = cloneDeep(matches);
    } else {
      renderedContext = cloneDeep(routerContext);
      renderedContext.__store.state.matches = [
        ...matches.map((m, i) => ({
          ...(prevMatches.current[i] || m),
          id: m.id,
        })),
        ...prevMatches.current.slice(matches.length),
      ];
    }

    return (
      <motion.div ref={ref} className="outlet" custom={direction} {...TransitionProps} {...props}>
        <RouterContext.Provider value={renderedContext}>
          <Outlet />
        </RouterContext.Provider>
      </motion.div>
    );
  }
);

export default AnimatedOutlet;

```

**Reply by scarabcoder** [reply]
Is the idea that the `<Outlet />` can just be swapped for `<AnimatedOutlet />`? It doesn't seem to be working out of the box for me, but also I'm not super familiar with framer-motion.

**Reply by fauxparse** [reply]
@scarabcoder No, you'll need to wrap it in `<AnimatePresence>`, as in the first example.

**Reply by scarabcoder** [reply]
@fauxparse Ahh, of course. It's working now, thank you!

**Reply by scarabcoder** [reply]
It seems to do some weird stuff when the router attempts to render a "404" route component, possibly because the route doesn't appear as a match but it is trying to render something different.

**Reply by aidrecabrera** [reply]
@fauxparse Nice! Where can I get the types for the `Direction` types?

**Reply by AndreiTS** [reply]
`type Direction = 'left' | 'right' | 'up' | 'down';`

**Reply by mjh316** [reply]
To catch the 404 page, you can do a `if (!nextMatch)` and then return a 404 component.

**Reply by JoshGlazebrook** [reply]
Using the above as-is, I keep seeing `Warning: Encountered two children with the same key, ``. Keys should be unique so that components maintain their identity across updates. Non-unique keys may cause children to be duplicated and/or omitted — the behavior is unsupported and could change in a future version.` in my console. Is this expected?

**Reply by CamBurris** [reply]
When I tried implementing this, I was getting this error: `TypeError: Cannot read private member #defaultOptions from an object whose class did not declare it`. This was happening because I was putting a TanStack Query client into the router context. The clone of the query client causes issues. I fixed it by cloning the routerContext and then manually assigning `options.context` back to the previous `options.context`. 

Code:
```typescript
if (isPresent) {
  const clone = cloneDeep(routerContext);
  clone.options.context = routerContext.options.context;

  renderedContext.current = clone;
}
```

**Reply by CamBurris** [reply]
Anyone else have issues with `<Link />`/`useLinkProps()` with this solution? When you pass in a `from` prop to either, clicking the link will throw an error saying `Invariant failed: Could not find match for from`. Here is a sandbox reproducing the issue: https://stackblitz.com/edit/tanstack-router-wdrreu. Go to the About page and then click the Test link.

**Reply by TheOrdinaryWow** [reply]
> ```ts
> <AnimatePresence mode="popLayout">
>         <AnimatedOutlet key={nextMatch.id} />
>       </AnimatePresence>
> ```

Just got same error here. It seems that 404 is not caught.

Got solved by modifying:

```tsx
<AnimatedOutlet key={nextMatch ? nextMatch.id : ""} />
```

**Reply by imSamiul** [reply]
> Anyone else have issues with `<Link />`/`useLinkProps()` with this solution? When you pass in a `from` prop to either, clicking the link will throw an error saying `Invariant failed: Could not find match for from`. Here is a sandbox reproducing the issue: https://stackblitz.com/edit/tanstack-router-wdrreu. Go to the About page and then click the Test link.

@CamBurris did you find any solution? I am facing the same problem on the `<Link>` component. It can't find the match.

**Reply by lexwebb** [reply]
Did anyone find a resolution for this?

**Reply by vladyoslav** [reply]
> Anyone else have issues with `<Link />`/`useLinkProps()` with this solution? When you pass in a `from` prop to either, clicking the link will throw an error saying `Invariant failed: Could not find match for from`. Here is a sandbox reproducing the issue: https://stackblitz.com/edit/tanstack-router-wdrreu. Go to the About page and then click the Test link.

> Did anyone find a resolution for this?

This example https://stackblitz.com/edit/tanstack-router-4rmnyw87 works with 

```
"@tanstack/react-router": "1.81.9",
"@tanstack/router-devtools": "1.81.9",
"@tanstack/router-generator": "1.81.9",
"@tanstack/router-plugin": "1.81.9"
```

But doesn't work with

```
"@tanstack/react-router": "1.81.10",
"@tanstack/router-devtools": "1.81.10",
"@tanstack/router-generator": "1.81.9",
"@tanstack/router-plugin": "1.81.9"
```

Downgrading to the version `1.81.9` solves the problem, so the error is hidden between these versions.

**Reply by VolDr** [reply]
any solution?

**Reply by vladyoslav** [reply]
> any solution?

The easiest way is to use a custom Link component throughout the app that passes the `from` property

The solution is crude and needs some work. Haven't tested it with relative routing,

```ts
import { type ComponentPropsWithRef, type FC } from 'react'

import { type LinkComponent, createLink } from '@tanstack/react-router'
import { motion } from 'motion/react'

interface AnimatedLinkProps extends ComponentPropsWithRef<typeof motion.a> {}

const AnimatedLinkComponent: FC<AnimatedLinkProps> = ({ ref, ...props }) => {
  return <motion.a ref={ref} {...props} />
}

const CreatedLinkComponent = createLink(AnimatedLinkComponent)

export const AnimatedLink: LinkComponent<typeof AnimatedLinkComponent> = ({
  ...props
}) => {
  // @ts-ignore
  return <CreatedLinkComponent preload="viewport" {...props} from="/" /> // TODO: normal `from` prop
}
```

**Reply by trompx** [reply]
On my side, adding `from="\"` to tanstack router `<Link>` component was enough to get rid of the invariant error.

**Reply by rhinck** [reply]
@trompx Do you know why this works?

**Reply by airtonix** [reply]
I know this is low effort, but here's a more portable version : 

```tsx
import { getRouterContext, Outlet } from '@tanstack/react-router';
import { motion, useIsPresent } from 'framer-motion';
import { cloneDeep } from 'lodash-es';
import { forwardRef, useContext, useRef } from 'react';

import type { ComponentPropsWithoutRef } from 'react';

export const createAnimatedOutlet = (
  transitionProps: Pick<
    ComponentPropsWithoutRef<typeof motion.div>,
    'initial' | 'animate' | 'exit'
  >,
) => {
  return forwardRef<HTMLDivElement>((_, ref) => {
    const RouterContext = getRouterContext();

    const routerContext = useContext(RouterContext);

    const renderedContext = useRef(routerContext);

    const isPresent = useIsPresent();

    if (isPresent) {
      renderedContext.current = cloneDeep(routerContext);
    }

    return (
      <motion.div ref={ref} {...transitionProps}>
        <RouterContext.Provider value={renderedContext.current}>
          <Outlet />
        </RouterContext.Provider>
      </motion.div>
    );
  });
};
```

**Reply by ThePeppy** [reply]
> Okay, I think I've cracked this! The trick seems to be keeping hold of the old router context and using that when the component is on the way out. Thankfully, TSR exposes all the context stuff you need to make this work:好的，我想我已经破解了这个问题！技巧似乎是要保留旧的 router 上下文，并在组件即将退出时使用它。幸运的是，TSR 暴露了你需要使这可行的一切上下文信息：
> 
> ```ts
> const AnimatedOutlet = forwardRef<HTMLDivElement>((_, ref) => {
>   const RouterContext = getRouterContext();
> 
>   const routerContext = useContext(RouterContext);
> 
>   const renderedContext = useRef(routerContext);
> 
>   const isPresent = useIsPresent();
> 
>   if (isPresent) {
>     renderedContext.current = cloneDeep(routerContext);
>   }
> 
>   return (
>     <motion.div ref={ref} {...transitionProps}>
>       <RouterContext.Provider value={renderedContext.current}>
>         <Outlet />
>       </RouterContext.Provider>
>     </motion.div>
>   );
> });
> ```
> 
> Then in your parent component:然后在你的父组件中：
> 
> ```ts
> const Root = () => {
>   const matches = useMatches();
>   const match = useMatch({ strict: false });
>   const nextMatchIndex = matches.findIndex((d) => d.id === match.id) + 1;
>   const nextMatch = matches[nextMatchIndex];
> 
>   return (
>     <main>
>       <AnimatePresence mode="popLayout">
>         <AnimatedOutlet key={nextMatch.id} />
>       </AnimatePresence>
>     </main>
>   );
> };
> ```
> 
> I haven't tested this super extensively, but initial results seem promising.我没有对这个方法进行彻底的测试，但初步结果看起来很有希望。
> 
> Notes:  备注：
> 
> * I haven't included the `transitionProps` here, but it's your standard Framer Motion stuff like `initial`, `animate`, and `exit`. Do whatever.我没有在这里包含 `transitionProps` ，但它是你标准的 Framer Motion 内容，比如 `initial` 、 `animate` 和 `exit` 。随便你怎么做。
> * This uses `deepClone` from `lodash` because I was in a hurry and had `lodash` in the project anyway, but you can substitute any deep clone function. The important part is that it MUST clone `context.store.state`: just saving a reference to the old context won't cut it.这使用了 `deepClone` 来自 `lodash` ，因为我当时很匆忙，项目里已经有了 `lodash` ，但你可以用任何深度克隆函数替换。重要的是它必须克隆 `context.store.state` ：仅仅保存旧上下文的引用是不够的。
> * This probably won't work with nested (animated) routes without some extra finesse (but I might be wrong!)这可能在嵌套（动画）路由中不起作用，除非有一些额外的技巧（但我也可能错了！）
> 
> Would love to know if anyone uses this, and how you get on!很想知道是否有人使用这个，以及你们的体验如何！

it's work!! thank you share it

**Comment by GuyFromEkb** [comment]
I can't get it to work too.

**Reply by fauxparse** [reply]
@GuyFromEkb Did you try what I posted above? If so, I'd be curious to know where it's going wrong for you, as this may help me to make a more robust version.

**Comment by AndreiTS** [comment]


**Reply by fauxparse** [reply]
This will have the same problem, where the exiting component will switch to the new route.

**Reply by AndreiTS** [reply]
Yes, you're right. I'm using your solution now and it's working.

**Comment by CamBurris** [comment]
Here is a solution using react transition group:

```typescript jsx
export const AnimatedOutlet = () => {
  const matches = useMatches();
  const match = useMatch({ strict: false });
  const matchIndex = matches.findIndex((d) => d.id === match.id);
  const nextMatchIndex =
    matchIndex === matches.length - 1 ? matchIndex : matchIndex + 1;
  const nextMatch = matches[nextMatchIndex];

  const RouterContext = getRouterContext();
  const routerContext = useContext(RouterContext);
  const renderedContext = useRef(routerContext);
  const isPresent = useRef(true);

  if (isPresent.current) {
    const clone = cloneDeep(routerContext);
    clone.options.context = routerContext.options.context;
    renderedContext.current = clone;
  }

  return (
    <FadeTransition
      mountOnEnter={true}
      unmountOnExit={true}
      appear={true}
      transitionKey={nextMatch.pathname}
      onEnter={() => {
        isPresent.current = true;
      }}
      onExit={() => {
        isPresent.current = false;
      }}
    >
      <RouterContextProvider router={renderedContext.current}>
        <Outlet />
      </RouterContextProvider>
    </FadeTransition>
  );
};
```

```typescript jsx
export const FadeTransition = ({
  transitionKey,
  children,
  ...props
}: Props) => {
  const nodeRef = useRef(null);

  return (
    <SwitchTransition>
      <CSSTransition
        {...props}
        key={transitionKey}
        timeout={150}
        nodeRef={nodeRef}
      >
        <TransitionCont ref={nodeRef}>{children}</TransitionCont>
      </CSSTransition>
    </SwitchTransition>
  );
};
```

**Reply by mwkcoding** [reply]
What is `TransitionCont`?

**Comment by dy0gu** [comment]
Is the accepted answer still the simplest away to achieve page transitions with Framer Motion? Seems like a big chore to have to manually implement simultaneous tracking of both the current and previous router context just to get transitions working. Every experience with TanStack Router has been vastly superior to React Router except for this little corner.

**Comment by brokeboiflex** [comment]
There should be an official guide on how to do it

**Comment by danielytics** [comment]
I was having a bunch of issues with the cloneDeep solution above:
1. [The context issue](https://github.com/TanStack/router/discussions/823#discussioncomment-10348993), this can be worked around as in the comment
2. [The `from` issue](https://github.com/TanStack/router/discussions/823#discussioncomment-10352521) has been ongoing for over half a year
3. I didn't like the implications of having to deep clone the router state on every page change

So I tried a different approach:

When the route changes, I capture the DOM node of the old page. Then I reinsert it into a new `motion.div` which I animate (initial to animate -- no need for AnimatePresence when we've already removed the node from the normal React component hierarchy) , and let React create a new node for the outlet. The key insight here is that animating out is purely visual, so we only need to capture the visual state -- which capturing the DOM node will do. Its not the prettiest solution, but it seems to work.

EDIT: Ok, so of course after posting, I found some issues. The updated code works, but its less efficient than I'd like because it clones the DOM nodes, which for complex trees could be more expensive than just deep cloning the router state. I noticed this is only an issue for my root level outlet, but I don't have time to dig into it more right now. So I've made cloning an option - by default it does it and everything works, but if you set `clone={false}` then it doesn't. I think the non-cloning method might be broken though because in some situations React doesn't like that its not seeing the nodes it expects. So turn off at your own risk.

Here it is in case its useful to others. I'm using tailwind, you may need to replace the classes to suit your own needs.

To use it, first wrap your root component with:

```
<AnimatedOutletWrapper>
  <App />
/AnimatedOutletWrapper>
```

And then replace your `<Outlet />` with something like this:

```typescript
<AnimatedOutlet
  from={Route.id}
  enter={{
    initial: { filter: 'blur(15px)', opacity: 0 },
    animate: { filter: 'blur(0)', opacity: 1 }
  }}
  exit={{
    initial: { filter: 'blur(0)', opacity: 1 },
    animate: { filter: 'blur(15px)', opacity: 0 }
  }}
  transition={{ duration: 0.3 }}
/>
```

WARNING: While I have tested it and it works for me so far, there may be some edge cases where it breaks. If you find any, please let me know!

Here's the implementation:

```typescript
import { useState, useEffect, useRef, createContext, useContext } from 'react';
import { useRouter, Outlet, AnyRoute, useMatch } from '@tanstack/react-router';
import { motion, MotionProps } from 'motion/react';

interface TransitionProps {
  initial?: MotionProps["initial"];
  animate?: MotionProps["animate"];
}

interface AnimatedOutletProps {
  enter: TransitionProps;
  exit: TransitionProps;
  transition?: MotionProps["transition"];
  from: AnyRoute['id']
  clone?: boolean
}


interface AnimatedOutletWrapperProps {
  children: React.ReactNode
}

type TakeSnapshotFn = () => void

type Registry = Map<string, TakeSnapshotFn>

const AnimatedOutletContext = createContext<Registry>(new Map())

function isDescendant(pathname: string, destinationPath: string) {
  return pathname === '/' ||
    (destinationPath.startsWith(pathname) &&
      (destinationPath.length === pathname.length ||
        destinationPath.charAt(pathname.length) === '/'))
}

export function AnimatedOutletWrapper({ children }: AnimatedOutletWrapperProps) {
  const router = useRouter();
  const registry = useRef<Registry>(new Map())

  useEffect(() => {
    // NOTE: This should be onBeforeNavigate, but due to https://github.com/TanStack/router/issues/3920 it's not working.
    // For now, we use onBeforeLoad, which runs right after onBeforeNavigate.
    // See: https://github.com/TanStack/router/blob/f8015e7629307499d4d6245077ad84145b6064a7/packages/router-core/src/router.ts#L2027
    const unsubscribe = router.subscribe('onBeforeLoad', ({ toLocation, pathChanged }) => {
      if (pathChanged) {
        const destinationPath = toLocation.pathname
        // Find the outlet with the longest pathname, that is part of the destination route
        let takeSnapshot: TakeSnapshotFn | null = null
        let longestLength = 0
        for (const [pathname, snapshotFn] of registry.current.entries()) {
          if (isDescendant(pathname, destinationPath) && pathname.length > longestLength) {
            longestLength = pathname.length
            takeSnapshot = snapshotFn
          }
        }
        if (takeSnapshot) {
          // Take a snapshot of the deepest outlet
          takeSnapshot()
        }
      }
    });
    return () => unsubscribe();
  }, [router]);

  return (
    <AnimatedOutletContext.Provider value={registry.current}>
      {children}
    </AnimatedOutletContext.Provider>
  )
}

export function AnimatedOutlet({
  enter,
  exit,
  transition = { duration: 0.3 },
  from,
  clone = true
}: AnimatedOutletProps) {
  const [snapshots, setSnapshots] = useState<{ node: HTMLElement, id: number }[]>([]);
  const [pathname, setPathname] = useState<string | null>(null)
  const outletRef = useRef<HTMLDivElement>(null);
  const nextId = useRef(0);

  const registry = useContext(AnimatedOutletContext)

  useEffect(() => {
    if (pathname) {
      registry.set(pathname, () => {
        const outletNode = outletRef.current!;
        const snapshotNode = outletNode.firstChild as HTMLElement;
        if (snapshotNode) {
          let node = snapshotNode;
          if (clone) {
            node = snapshotNode.cloneNode(true) as HTMLElement;
          } else {
            outletNode.removeChild(snapshotNode)
          }
          const newSnapshot = { node, id: nextId.current++ }
          setSnapshots((prevSnapshots) => [...prevSnapshots, newSnapshot]);
        }
      });

      return () => { registry.delete(pathname) };
    }
  }, [registry, pathname]);

  const handleAnimationComplete = (id: number) => {
    setSnapshots((prevSnapshots) =>
      prevSnapshots.filter((snapshot) => snapshot.id !== id)
    );
  };

  return (
    <>
      {pathname === null && <GetPathName setPathname={setPathname} from={from} />}
      <div className="relative w-full h-full">
        {snapshots.map((snapshot) => (
          <motion.div
            key={snapshot.id}
            className="absolute inset-0 pointer-events-none w-full h-full"
            initial={exit.initial}
            animate={exit.animate}
            transition={transition}
            onAnimationComplete={() => handleAnimationComplete(snapshot.id)}
            aria-hidden="true"
            ref={(el) => {
              if (el && snapshot.node) {
                el.appendChild(snapshot.node);
              }
            }}
          />
        ))}

        <motion.div
          key={nextId.current}
          ref={outletRef}
          className="relative w-full h-full"
          initial={enter.initial}
          animate={enter.animate}
          transition={transition}
        >
          <div className="w-full h-full">
            <Outlet />
          </div>
        </motion.div>
      </div>
    </>
  );
}

function GetPathName({ from, setPathname }: { from: AnyRoute['id'], setPathname: (pathname: string) => void }) {
  const match = useMatch({ from })
  useEffect(() => {
    setPathname(match.pathname)
  }, [match.pathname])
  return null
}
```

A few things to note:
1. The wrapper and context is used so that nested AnimatedOutlets work. If you only have one (eg root level) animated outlet and don't need animation for sub routes, then this could be simplified by moving the router.subscribe() into the AnimatedOutlet itself. I actually started with that, until I noticed nested routes broke.
2. This version tracks multiple snapshots. This is so that if you navigate faster than the animations complete, it will continue animating previous animations. Without it, any older animations would stop abruptly and the component disappear. If you don't need this, the logic could be simplified to store only a single snapshot per outlet instead of maintaining a list.
3. The GetPathName component could be removed and useMatch called directly in AnimatedOutlet. Its here only to prevent AnimatedOutlet components from rendering on every route change, even if they don't need to.

I'm not sure if this is useful to anyone. Besides fixing the bug (2 in the list), its far more complex than the other solutions presented... but maybe someone who understands React internals better can use this as a starting point and build on it.

**Comment by yaindrop** [comment]
Having read all the above discussion I find none of them quite satisfactory... 

I guess as of now the best we can do is to avoid using `<Outlet />` and `{ component: <SomePage /> }` altogether and just use the route matches as simple global states, managing view appearance manually in the traditional way.

This gives us full control of transition and other behaviors while still enjoying the benifit of history management / statically typed route tree of the Tanstack Router.

**Reply by danielytics** [reply]
I actually abandoned my previous and went with with a page wrapper setup which animates out completely before actually changing the route. It does mean your entry and exit animations can’t overlap but it’s been a lot more robust than the previous solutions.

However, react-experimental includes view transition support that looks very good, so I think tanstack routers view transition based animation will become a viable built in alternative when it lands. It still won’t allow you to use Motion to animate, but at least it’ll be react built in functionality. My primary reason for wanting to use Motion is that view transitions have been really awkward and clunky to use, but that will be changing.

**Comment by marek-hanzal** [comment]
Hello, I would like to add little piece which may inspire somebody here:

```tsx
import { useRouterState } from "@tanstack/react-router";
import { AnimatePresence, motion } from "motion/react";
import {
	type FC,
	type PropsWithChildren,
	useLayoutEffect,
	useRef,
	useState,
} from "react";

export const PageTransition: FC<PropsWithChildren> = ({ children }) => {
	const { pathname, status } = useRouterState({
		select: (state) => ({
			pathname: state.location.pathname,
			status: state.status,
		}),
	});
	const [transition, setTransition] = useState<"ghost" | "outlet">("outlet");

	const outletRef = useRef<HTMLDivElement>(null);
	const ghostRef = useRef<string | undefined>(undefined);
	const readyRef = useRef(false);

	useLayoutEffect(() => {
		if (status === "idle") {
			readyRef.current = true;
		}
		if (status === "pending" && readyRef.current) {
			ghostRef.current = outletRef.current?.innerHTML;
			setTransition("ghost");
		}
	}, [
		status,
	]);

	return (
		<AnimatePresence
			mode="wait"
			initial={false}
		>
			{transition === "ghost" ? (
				<motion.div
					key={`ghost:${pathname}`}
					initial={{
						opacity: 1,
					}}
					animate={{
						opacity: 1,
					}}
					exit={{
						opacity: 0,
						scale: 0.75,
					}}
					transition={{
						duration: 0.25,
						ease: "easeInOut",
					}}
					onAnimationComplete={() => {
						setTransition("outlet");
						ghostRef.current = undefined;
					}}
					dangerouslySetInnerHTML={{
						__html: ghostRef.current ?? "",
					}}
					className="w-full h-full"
				/>
			) : null}

			{transition === "outlet" ? (
				<motion.div
					ref={outletRef}
					key={`${status}-${pathname}`}
					initial={{
						opacity: 0,
						rotateZ: 30,
						scale: 0.75,
					}}
					animate={{
						opacity: 1,
						x: 0,
						y: 0,
						scale: 1,
						rotateY: 0,
						rotateX: 0,
						rotateZ: 0,
					}}
					transition={{
						duration: 0.25,
						ease: "easeInOut",
					}}
					className="w-full h-full"
				>
					{children}
				</motion.div>
			) : null}
		</AnimatePresence>
	);
};
```

- it's using TanStack Router, but it should be basically portable to (almost?) anything
- the core hack is to take innerHtml and use it as a ghost layer when animated
- "children" here could be whatever you want, keep in mind styles needs some attention, but you're clever guys, you know what to do :)
- current animations are crazy, I've used them only to mainly showcase, if the stuff works properly
- by ChatGPT opinion, forms won't work as expected - you've to do a bit more magic during "innerHtml" step
- You've also try the limitations - I've used it on fairly simple pages, dunno what happens e.g. on big grids and so on
- The flow (in TSR): wait until new route is loaded (start on pending) and execute animations to exchange previous content with the new one
- *Presence::mode = "wait" - important unless you update styles to overlay both animations or they'll render under each other (...just ugly stuff)

And that's it - I hope somebody will be happy with this and it will serve you.

See ya

**Comment by klapacz** [comment]
Hey, I wrote a [blog post](https://klapacz.dev/blog/0003-tanstack-router-animated-modals/) on using View Transitions to animate page transitions in TanStack Router. Hope it helps someone.

**Comment by nukeop** [comment]
So this approach is completely broken in the new release. The "minor" version actually makes this completely unfeasible because of massive architectural changes. You can't `cloneDeep` the router context anymore.

**Reply by scarabcoder** [reply]
Yeah, I noticed this broke recently as well. The `RouterContext` is no longer exported in the newer versions. Real shame, because it worked really well for us.

**Reply by nukeop** [reply]
I'm experimenting with cloning via Proxy and then using that:

```typescript
const createFrozenRouter = (router: AnyRouter): AnyRouter => {
  const frozenState = cloneDeep(router.state);
  const frozenStore = {
    state: frozenState,
    get: () => frozenState,
    subscribe: () => ({ unsubscribe: () => {} }),
  };

  return new Proxy(router, {
    get(target, prop, receiver) {
      if (prop === '__store') return frozenStore;
      if (prop === 'state') return frozenState;
      return Reflect.get(target, prop, receiver);
    },
  });
};
```

But this ends up being janky and the animations aren't as smooth.

**Comment by nukeop** [comment]
Somebody figured it out here:

https://github.com/libnyanpasu/clash-nyanpasu/blob/main/frontend/nyanpasu/src/components/router/animated-outlet.tsx

This is an animated outlet that works with the newly updated tanstack router. The new pattern is pretty different. Instead of cloning the router context, they create a fake router with a frozen __store that returns a snapshot of the old state.

**Reply by pklepa** [reply]
This indeed works but jesus... One would hope a reasonably common feature as page transitions wouldn't require all this to work
